src/ray.rs

```diff
@@ -103,22 +87,118 @@ fn min(x: f32, y: f32) -> f32 {
 /// max( 1.0,  NaN):  NaN
 /// ```
 #[inline(always)]
-fn max(x: f32, y: f32) -> f32 {
+#[allow(dead_code)]
+fn max<T: Scalar + Copy + PartialOrd>(x: T, y: T) -> T {
     if x > y {
         x
     } else {
         y
     }
 }
 
-impl Ray {
+trait IntersectsAABB<X, T, const D: usize>
+where
+    T: Scalar + Copy + ClosedSub + ClosedMul + Zero + PartialOrd + SimdPartialOrd,
+{
+    fn ray_intersects_aabb(&self, aabb: &AABB<T, D>) -> bool;
+}
+
+impl<T, const D: usize> IntersectsAABB<Ray<T, D>, T, D> for Ray<T, D>
+where
+    T: Scalar + Copy + ClosedSub + ClosedMul + Zero + PartialOrd + SimdPartialOrd,
+{
+    #[inline(always)]
+    default fn ray_intersects_aabb(&self, aabb: &AABB<T, D>) -> bool {
+        let lbr = (aabb[0].coords - self.origin.coords).component_mul(&self.inv_direction);
+        let rtr = (aabb[1].coords - self.origin.coords).component_mul(&self.inv_direction);
+
+        let (inf, sup) = lbr.inf_sup(&rtr);
+
+        let tmin = inf.max();
+        let tmax = sup.min();
+
+        tmax > max(tmin, T::zero())
+    }
+}
+
+#[inline(always)]
+#[cfg(target_arch = "x86_64")]
+fn vec_to_mm(vec: &SVector<f32, 3>) -> __m128 {
+    unsafe { _mm_set_ps(vec.z, vec.z, vec.y, vec.x) }
```

Comment:
Really digging all this custom SIMD but I'm wondering whether perhaps [packed_simd](https://github.com/rust-lang/packed_simd) might be the way forward?

Reply:
packed_simd says on the readme that std::simd is the way forward

Reply:
Hm true. I didn't know there's still no good stabilized SIMD stuff in Rust. What do you think instead about [wide](https://github.com/Lokathor/wide)?

Reply:
I'll look into it and maybe trying Simba (what nalgebra uses under the hood) again. But with Simba I had performance magnitudes worse that with the pure simd, and I'm unsure why

Reply:
There's also [simdeez](https://crates.io/crates/simdeez) which might be worth checking out.